src/hotspot/share/utilities/globalDefinitions_gcc.hpp

```diff
@@ -139,10 +139,21 @@ inline int g_isfinite(jdouble f)                 { return isfinite(f); }
 #endif // _LP64
 
 // gcc warns about applying offsetof() to non-POD object or calculating
-// offset directly when base address is NULL. Use 16 to get around the
-// warning. The -Wno-invalid-offsetof option could be used to suppress
-// this warning, but we instead just avoid the use of offsetof().
-#define offset_of(klass,field) (size_t)((intx)&(((klass*)16)->field) - 16)
+// offset directly when base address is NULL. The -Wno-invalid-offsetof
+// option could be used to suppress this warning, but we instead just
+// avoid the use of offsetof().
+//
+// FIXME: This macro is complex and rather arcane. Perhaps we should
+// use offsetof() instead, with the invalid-offsetof warning
+// temporarily disabled.
+#define offset_of(klass,field)                  \
+[]() {                                          \
+  char space[sizeof (klass)];                   \
+  klass* dummyObj = (klass*)space;              \
+  char* c = (char*)(void*)&dummyObj->field;     \
+  return (size_t)(c - space);                   \
+}()
```

Comment:
I am very surprised this does not produce any warnings! Let me test this on different GCCs...

Reply:
Of course we can't exclude the possibility of warnings, given that taking the offset of a field of a non-POD class is unspecified behaviour. Having said that, I'm not sure exactly what it'd be warning about.

Reply:
This restriction in C++ is unnecessary, really, since AFAIK this works on all implementations.

Reply:
This seems to work on GCCs 9..12 in {x86_64, x86_32, aarch64}.

Reply:
FYI: I have verified this also works with GCC version 11.3.0 on linux-riscv64 platform.

Reply:
Thanks.

Reply:
How does this behave with respect to alignment?
The previous implementation used 16 which is at least aligned to `std::max_align_t` (for all hotspots build systems?) 
My understanding is that `char space[sizeof (klass)];` is 1-byte aligned, so `dummyObj` may be misaligned.

Reply:
> How does this behave with respect to alignment? The previous implementation used 16 which is at least aligned to `std::max_align_t` (for all hotspots build systems?) My understanding is that `char space[sizeof (klass)];` is 1-byte aligned, so `dummyObj` may be misaligned.

That's an interesting point. I don't propose to trawl the standard for exact wording, but as far as I recall it's _accesses_ to misaligned members that are problematic.

Reply:
I had wondered the same thing, and had already spent some time trawling the
standard. I think the only way to get a misaligned pointer (without using
`#pragma pack`) involves casts. The relevant `reinterpret_cast` is defined as
being equivalent to `static_cast` to `void*` then `static_cast` to `T*` (C++14
5.2.10/7). There it says round trip cast is okay with same alignment
requirements, but is silent about when they are different. However,
`static_cast` from `void*` to `T*` is okay if the address satisfies the
alignment requirement for `T`, else unspecified (C++14 5.2.9/13).

So if the space buffer is misaligned for the klass type, the cast will have an
unspecified result. There may also be a question regarding whether
`&dummyObj->field` constitutes an access.  I recall a gcc bug with a long
discussion about whether `&x[n]` constituted an access (with gcc maintainers
saying yes), but I couldn't refind that bug today.

I think the safe thing to do is to align the space buffer.  The simplest way
to do that would be with `alignas`, but using that is pending approval:
https://git.openjdk.org/jdk/pull/11446. `ATTRIBUTE_ALIGNED` is the HotSpot
macro for that.